source/driverHandler.py

```diff
@@ -0,0 +1,234 @@
+# -*- coding: UTF-8 -*-
+#driverHandler.py
+#A part of NonVisual Desktop Access (NVDA)
+#This file is covered by the GNU General Public License.
+#See the file COPYING for more details.
+#Copyright (C) 2006-2018 NV Access Limited, Leonard de Ruijter
+
+"""Handler for driver functionality that is global to synthesizers and braille displays."""
+
+from baseObject import AutoPropertyObject
+import config
+from copy import deepcopy
+
+class Driver(AutoPropertyObject):
+	"""
+	Abstract base class for drivers, such as speech synthesizer and braille display drivers.
+	Abstract subclasses such as L{braille.BrailleDisplayDriver} should set L{_configSection}.
+
+	At a minimum, drivers must set L{name} and L{description} and override the L{check} method.
+
+	L{supportedSettings} should be set as appropriate for the settings supported by the driver.
+	Each setting is retrieved and set using attributes named after the setting;
+	e.g. the L{dotFirmness} attribute is used for the L{dotFirmness} setting.
+	These will usually be properties.
+	"""
+
+	#: The name of the driver; must be the original module file name.
+	#: @type: str
+	name = ""
+	#: A description of the driver.
+	#: @type: str
+	description = ""
+	#: The configuration section where driver specific subsections should be saved.
+	#: @type: str
+	_configSection = ""
+
+	def __init__(self):
+		"""Initialize this driver.
+		This method can also set default settings for the driver.
+		@raise Exception: If an error occurs.
+		@postcondition: This driver can be used.
+		"""
+		super(Driver, self).__init__()
+
+	def terminate(self):
+		"""Terminate this driver.
+		This should be used for any required clean up.
+		@precondition: L{initialize} has been called.
+		@postcondition: This driver can no longer be used.
+		"""
+
+	_abstract_supportedSettings = True
+	def _get_supportedSettings(self):
+		"""The settings supported by the driver.
+		@rtype: list or tuple of L{DriverSetting}
+		"""
+		return ()
+
+	@classmethod
+	def check(cls):
+		"""Determine whether this driver is available.
+		The driver will be excluded from the list of available drivers if this method returns C{False}.
+		For example, if a speech synthesizer requires installation and it is not installed, C{False} should be returned.
+		@return: C{True} if this driver is available, C{False} if not.
+		@rtype: bool
+		"""
+		return False
+
+	def isSupported(self,settingName):
+		"""Checks whether given setting is supported by the driver.
+		@rtype: l{bool}
+		"""
+		for s in self.supportedSettings:
+			if s.name==settingName: return True

```

Comment:
How does this work with translations? We shouldn't be comparing translated strings.
Oh, I see there is also displayName. I would have prefered name be ID I think.
But then StringParameterInfo has ID and name where name is the displayable string.

Reply:
Mm, yes, this is confusing. I'd be happy to change name to id, that's not a big deal. however, I want to use self.id, not self.ID. So I would like to propose making things uniform here. How about self.id and self.displayName for both settings and parameters?

Reply:
> How about self.id and self.displayName for both settings and parameters?

Yes, that is fine with me.